dipy/segment/mrf.pyx

```diff
@@ -0,0 +1,657 @@
+    Image segmentation by maximum a-posteriori estimation of HMRF model
+    with the Potts/Ising potential
+    """
+
+    def __init__(self):
+        
+        r""" Initializes an instance of the ConstantObservationModel class
+        """
+        pass
+
+
+    def initialize_param_uniform(self, image, nclasses):
+        
+        r""" Initializes the means and variances uniformly
+
+        The means are initialized uniformly along the dynamic range of `image`.
+        The variances are set to 1 for all classes
+
+        Parameters
+        ----------
+        image : array, shape(X, Y, Z)
+        nclasses : int, number of desired classes
+
+        Returns
+        -------
+        mu : array, shape(nclasses,)
+        sigmasq : array, shape(nlasses,)
+        """
+
+        cdef:
+            double[:] mu = np.empty((nclasses,), dtype=np.float64)
+            double[:] sigmasq = np.empty((nclasses,), dtype=np.float64)
+
+        _initialize_param_uniform(image, mu, sigmasq)
+        return np.array(mu), np.array(sigmasq)
+
+
+    def seg_stats(self, input_image, seg_image, nclass):
+        
+        r""" Mean and standard variation for N desired  tissue classes
+
+        Parameters
+        ----------
+        input_image : ndarray, gray level T1 image
+        seg_image : ndarray, initital segmentation image
+        nclass : int, number of classes (3 in most cases)
+
+        Returns
+        -------
+        mu, std, var : ndarrays of dimensions 1x3
+                        Mean, standard deviation and variance for every class
+
+        """
+        mu = np.zeros(nclass)
+        std = np.zeros(nclass)
+        var = np.zeros(nclass)
+
+        for i in range(0, nclass):
+
+            H = input_image[seg_image == i]
+
+            mu[i] = np.mean(H, -1)
+            std[i] = np.std(H, -1)
+            var[i] = np.var(H, -1)
+
+        return mu, std, var
+
+
+    def negloglikelihood(self, image, mu, sigmasq, nclasses):
+        
+        r""" Computes the gaussian negative log-likelihood of each class at 
+        each voxel of `image` assuming a gaussian distribution with means and
+        variances given by `mu` and `sigmasq`, respectively (constant models
+        along the full volume). The negative log-likelihood will be written
+        in `nloglike`.
+        
+        Parameters
+        ----------
+        image : ndarray, gray scale T1 image
+        mu : ndarray, mean of each class
+        sigmasq : ndarray, variance of each class
+        nclasses : int, number of classes
+        
+        Returns
+        -------
+        nloglike : 4D ndarray, negloglikelihood for each class in each volume  
+
+        """
+
+        nloglike = np.zeros(image.shape + (nclasses,), dtype=np.float64)
+
+        for l in range(nclasses):
+            _negloglikelihood(image, mu, sigmasq, l, nloglike)
+
+        return nloglike
+
+
+    def prob_neighborhood(self, image, seg, beta, nclasses):

```

Comment:
Hi Julio =)
I totally agree, `prob_neighborhood` requires `beta`, but `beta` is not a parameter of the observation model. What I mean is that the class `ConstantObservationModel` is only responsible for computing things related to the likelihood (in this particular case, it only requires `mu` and `sigma` for each tissue type). `beta` is a parameter of the prior (in this case, the prior is that the label field is a MRF with Ising potential and parameter `beta`). The problem is that for a given observation model, we may have many different priors, and a prior can be used with many different observation models (e.g. using splines instead of constant models) that's why we should decouple the likelihood and the prior.

ICM is in charge of maximizing the posterior distribution with respect to the segmentation (with a fixed log-likelihood, which was provided by the observation model). Therefore, ICM doesn't need to know any detail about the model, it doesn't care about the mu's and sigma's, it just needs likelihoods, which may have been computed assuming constant models, linear models, splines, etc. 

I hope it helps.